app/server/MergedServer.ts

```diff
@@ -0,0 +1,234 @@
+  externalStorage?: boolean;
+}
+
+export class MergedServer {
+
+  public static async create(port: number, serverTypes: ServerType[], options: ServerOptions = {}) {
+    options.settings ??= await getGlobalConfig();
+    const ms = new MergedServer(port, serverTypes, options);
+    // We need to know early on whether we will be serving plugins or not.
+    if (ms.hasComponent("home")) {
+      const userPort = checkUserContentPort();
+      ms.flexServer.setServesPlugins(userPort !== undefined);
+    } else {
+      ms.flexServer.setServesPlugins(false);
+    }
+
+    ms.flexServer.addCleanup();
+    ms.flexServer.setDirectory();
+
+    if (process.env.GRIST_TEST_ROUTER) {
+      // Add a mock api for adding/removing doc workers from load balancer.
+      ms.flexServer.testAddRouter();
+    }
+
+    if (ms._options.logToConsole !== false) { ms.flexServer.addLogging(); }
+    if (ms._options.externalStorage === false) { ms.flexServer.disableExternalStorage(); }
+    await ms.flexServer.addLoginMiddleware();
+
+    if (ms.hasComponent("docs")) {
+      // It is important that /dw and /v prefixes are accepted (if present) by health check
+      // in ms case, since they are included in the url registered for the doc worker.
+      ms.flexServer.stripDocWorkerIdPathPrefixIfPresent();
+      ms.flexServer.addTagChecker();
+    }
+
+    ms.flexServer.addHealthCheck();
+    if (ms.hasComponent("home") || ms.hasComponent("app")) {
+      ms.flexServer.addBootPage();
+    }
+    ms.flexServer.denyRequestsIfNotReady();
+
+    if (ms.hasComponent("home") || ms.hasComponent("static") || ms.hasComponent("app")) {
+      ms.flexServer.setDirectory();
+    }
+
+    if (ms.hasComponent("home") || ms.hasComponent("static")) {
+      ms.flexServer.addStaticAndBowerDirectories();
+    }
+
+    await ms.flexServer.initHomeDBManager();
+    ms.flexServer.addHosts();
+
+    ms.flexServer.addDocWorkerMap();
+
+    if (ms.hasComponent("home") || ms.hasComponent("static")) {
+      await ms.flexServer.addAssetsForPlugins();
+    }
+
+    if (ms.hasComponent("home")) {
+      ms.flexServer.addEarlyWebhooks();
+    }
+
+    if (ms.hasComponent("home") || ms.hasComponent("docs") || ms.hasComponent("app")) {
+      ms.flexServer.addSessions();
+    }
+
+    ms.flexServer.addAccessMiddleware();
+    ms.flexServer.addApiMiddleware();
+    await ms.flexServer.addBillingMiddleware();
+
+    return ms;
+  }
+
+  public readonly flexServer: FlexServer;
+  private readonly _serverTypes: ServerType[];
+  private readonly _options: ServerOptions;
+
+  private constructor(port: number, serverTypes: ServerType[], options: ServerOptions = {}) {
+    this._serverTypes = serverTypes;
+    this._options = options;
+    this.flexServer = new FlexServer(port, `server(${serverTypes.join(",")})`, options);
+  }
+
+  public hasComponent(serverType: ServerType) {
+    return this._serverTypes.includes(serverType);
+  }
+
+
+  public async run() {
```

Comment:
This is the second part of the old `main` function.

The FlexServer is created in `create()`, but before calling `run()` it is not started. Between these two function calls, one can retrieve and modify certain FlexServer components.